pkg/proxy/winkernel/proxier.go

```diff
@@ -1302,6 +1321,28 @@ func (proxier *Proxier) syncProxyRules() {
 			}
 			lbIngressIP.hnsID = hnsLoadBalancer.hnsID
 			klog.V(3).InfoS("Hns LoadBalancer resource created for loadBalancer Ingress resources", "lbIngressIP", lbIngressIP)
+
+			if proxier.forwardHealthCheckVip && gatewayHnsendpoint != nil {
+				nodeport := proxier.healthzPort
```

Comment:
but this is not correct, this is about the healthz of the apiserver, not about if the service has local endpoints ... but to be honest I don't fully understand the problem, what is the TL; DR, of why the NewServiceHealthServer doesn't work in windows? that should fix the problem, no?
```
	serviceHealthServer := healthcheck.NewServiceHealthServer(hostname, recorder, []string{} /* windows listen to all node addresses */)
```

Reply:
The problem is it's not the node address. The google load balancer works this way:

```
User --> [clientip:12345 to lb_vip:80] -> Google LB -> [clientip:12345 to lb_vip:80] -> Node -> [clientip:12345 to pod_ip:80] -> Pod 
```
This works right now in kubeproxy's winkernel mode, because the rules are already in place.

What doesn't work is the health check for normal service:
```
Google LB -> [lb:12345 to lb_vip:10256] -> Node -> X(dropped)
```
and for `externalTrafficPolicy: local` service
```
Google LB -> [lb:12345 to lb_vip:<per-service-hc-port>] -> Node -> X(dropped)
```
They are dropped because the destination is not among any node IPs.

This basically adds hns rules to change request to node IP.

Reply:
yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

Reply:
the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this :sweat_smile:

Reply:
> yeah, but why it uses the kube-proxy healthz port , it should use the healthzcheck nodeport, no?

`proxier.healthzPort` is basically 10256 if it's not explicitly configured: 

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L212-L220

https://github.com/kubernetes/kubernetes/blob/1367cca8fd67b09606b01c0a9e46cef59aef3424/cmd/kube-proxy/app/server.go#L226

https://github.com/kubernetes/kubernetes/blob/1bd00776b5d78828a065b5c21e7003accc308a06/pkg/cluster/ports/ports.go#L36

this is basically the healthzcheck nodeport. For `externalTrafficPolicy: local`, the port is different for every service. 

> the part I'm lost is about when it has to use healthz port, but I'm not familiar with the internals of GCE LBs, maybe we should get a reviewer more familiar with this 😅

I understand it's really not straight-forward.. I'll try to get someone from GKE service team to review as well :) @robscott @bowei can you guys help take a look?

Reply:
Just catching up on this thread. I _think_ this makes sense. Although we don't need to do this in other proxier implementations, it makes sense that we'd need to this lb_vip -> node ip translation here to get health checks working. I'm less familiar with how the primary/default health check needs to work here. Regardless of OS, that seems to be handled exclusively at a higher level, with no extra logic added in either iptables or winkernel proxier implementations:

https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_others.go#L153
https://github.com/kubernetes/kubernetes/blob/481459d12dc82ab88e413886e2130c2a5e4a8ec4/cmd/kube-proxy/app/server_windows.go#L101

I'm not familiar enough with how this is setup to know what's necessary, but is it possible that this works without any changes for the default health check? The original bug as reported by the PR seems to indicate the problem is only with `ExternalTrafficPolicy: Local`.

Reply:
at higher level, `healthcheck.NewProxierHealthServer()` is called to create http server that listens on the specific bind address for health check purpose. Because it listens on `0.0.0.0`, on Linux it will accept the connection regardless of destination. Unfortunately for Windows, the additional forward needs to be done to forward VIP to node-ip so the health check server can receive traffic.